File: chapter01/URLify.js
function URLify(s) {
  let word = "";
  let sentence = [];

  for (let i = 0; i < s.length; i++) {
    if (s[i] !== " ") {
      word += s[i];

      if (i + 1 === s.length) {
        sentence.push(word);
      }

    } else {
      if (word.length > 0) {
        sentence.push(word);
      }

      word = "";
    }
  }

  return sentence.join("%20");
}

console.log(URLify("Mr  John  Smith    ")); // "Mr%20John%20Smith"
console.log(URLify("   Mr John      Smith")); // "Mr%20John%20Smith"

File: chapter01/URLify/URLify.java
import java.util.*;

public class URLify {
    public static String solution(String url, int trueLength) {
        char[] urlArray = url.toCharArray();
        int spaceCount = 0;

        for (int i = 0; i < trueLength; i++) {
            if (urlArray[i] == ' ') {
                spaceCount++;
            }
        }

        int newLength = trueLength + spaceCount * 2;
        int index = newLength - 1;

        for (int i = trueLength - 1; i >= 0; i--) {
            if (urlArray[i] == ' ') {
                urlArray[index] = '0';
                urlArray[index - 1] = '2';
                urlArray[index - 2] = '%';

                index -= 3;
            } else {
                urlArray[index] = urlArray[i];

                index--;
            }
        }

        return new String(urlArray, 0, newLength);

    }

    public static void main(String[] args) {
        String url = "Mr John Smith     ";
        int trueLength = 13;

        System.out.println(solution(url, trueLength));
    }
}
